### ISSUE TITLE
[AliasAnalysis] getModRefInfo(DetachInst *, MemoryLocation &) unnecessarily superlinear time, unnecessarily recursive.

### ISSUE THREAD

**Issue by VictorYing** [wmoses/Parallel-IR#92]
Notice how this function calls some variant of getModRefInfo(I) for each non-sync instruction I contained within the detached CFG:

https://github.com/wsmoses/Tapir-LLVM/blob/ba2f0eda34108ab64e581f2740b4d06eecc50709/lib/Analysis/AliasAnalysis.cpp#L487-L526

This includes calling getModRefInfo() on nested detaches, making this function (mutually) recursive (with getModRefInfo)!  This is totally unnecessary: The (non-recursive) loop will already iterate through all instructions contained within the detached CFG region. Calling getModRefInfo() recursively on a nested detach will simply require redundantly iterating through all the instructions that are also within the nested detached CFG. If there are N levels of nested detaches, then the code within the innermost detach will be iterated through N times. In the worst case, calling getModRefInfo() on an outer detach can take quadratic time in the size of the detached region, and linear depth of recursive calls.

The right answer is probably to treat nested detaches just like nested syncs: they can just be ignored. 
getModRefInfo(DetachInst *) should be able to compute it's result with a single linear pass over the detached region, and never needs to recursively call itself.

If I'm right, then fixing this will have the nice benefit of removing a case of nasty recursion that can cause a stack overflow on inputs with particularly deeply nested task structures, which I've been running into in some of my autoparallelization efforts.

**Comment by VictorYing** [comment]
Here's a more recent version that still has the same problem:

https://github.com/wsmoses/Tapir-LLVM/blob/a038195c28ef6c41f5ad8e0947f061e71771cb6a/lib/Analysis/AliasAnalysis.cpp#L657-L694

It's that
```c++
Result = unionModRef(Result, getModRefInfo(&I, Loc)); 
```
that causes the problem without any checks for whether this I is itself a detach.

**Comment by VictorYing** [comment]
Elsewhere in the same file, there's a different loop that correctly ignores nested syncs and detaches within a detached CFG:
https://github.com/wsmoses/Tapir-LLVM/blob/a038195c28ef6c41f5ad8e0947f061e71771cb6a/lib/Analysis/AliasAnalysis.cpp#L145-L180
Note the
```c++
if (isa<SyncInst>(DI) || isa<DetachInst>(DI))
  continue; 
```

**Comment by VictorYing** [comment]
Closed by 83765ecbfc630e63c6a6b9d12c1ac1d969a691d6
